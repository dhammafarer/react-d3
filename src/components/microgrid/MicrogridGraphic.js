import React from 'react';
import PropTypes from 'prop-types';
import './MicrogridChart.scss';
import { isometricTileCoords, isometricTilePolygonPoints } from '../../helpers/isometric-grid.js';
import * as o from '../../data/microgrid-objects.js';
import GroundTiles from './GroundTiles.js';
import BuildingTiles from './BuildingTiles.js';
import IsometricGrid from './IsometricGrid.js';
import BuildingMarkers from './BuildingMarkers.js';

const tileRatio = 1.7345;

class MicrogridGraphic extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      size: {
        width: 400,
        height: 400 * tileRatio,
        margin: [0, 0]
      },
      gridSize: [3, 3],
      tile: {
        width: 200,
        height: 200 * tileRatio
      },
      groundMap: [
        [o.grass, o.grass, o.grass],
        [o.grass, o.grass, o.grass],
        [o.grass, o.grass, o.grass]

      ],
      buildingsMap: [
        [o.temple, null, o.solar],
        [o.house, null, o.house],
        [null, o.factory, null]
      ]
    };
  }
  componentDidMount () {
    this.setGraphicSize();
  }

  setGraphicSize () {
    let componentWidth = this.graphic.offsetWidth;
    let tileWidth = componentWidth / (this.state.gridSize[1] + 1);
    let tileHeight = tileWidth / tileRatio;
    let margin = [tileHeight / 2, tileWidth / 2];
    let width = componentWidth - margin[1] * 2;
    let height = tileHeight * (this.state.gridSize[1]);
    this.setState({
      size: {width, height, margin},
      tile: {width: tileWidth, height: tileHeight}
    });
  }

  mapToIso (objectsMap) {
    return objectsMap
      .map((row, y) => row
        .map((tile, x) => {
          let coords = this.tileCoords()[y][x];
          let style = {
            left: coords.x,
            top: coords.y,
            width: this.state.tile.width
          };
          return {tile, pos: [x, y], style};
        })
      )
      .reduce((a, b) => a.concat(b), []);
  }

  tileCoords () {
    let {width, height} = this.state.tile;
    return isometricTileCoords(this.state.gridSize, [width, height]);
  }

  tilePolygons () {
    let {width, height} = this.state.tile;
    return this.tileCoords()
      .map((row, y) => row.map((coords, x) => {
        return {
          points: isometricTilePolygonPoints([width, height], coords),
          pos: [y, x]};
      }))
      .reduce((a,b) => a.concat(b), []);
  }

  render () {
    let {width, height, margin} = this.state.size;
    let graphicStyles = {
      width, height,
      margin: margin[0] + 'px ' + margin[1] + 'px'
    };

    let ground = this.mapToIso(this.state.groundMap);
    let buildings = this.mapToIso(this.state.buildingsMap).filter(el => el.tile);
    let grid = this.tilePolygons();

    return (
      <div className="system-graphic" ref={graphic => this.graphic = graphic}>
        <div className="graphic-content" style={graphicStyles}>

          <GroundTiles ground={ground}/>
          <BuildingTiles buildings={buildings}/>
          <IsometricGrid grid={grid} width={width} height={height}/>
          <BuildingMarkers data={buildings} height={this.state.tile.height}/>

        </div>
      </div>
    );
  }
}

MicrogridGraphic.propTypes = {
  timeline: PropTypes.object
};

export default MicrogridGraphic;
