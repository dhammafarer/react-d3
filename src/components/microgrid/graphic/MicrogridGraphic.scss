.system-graphic {
}

.graphic-content {
  position: relative;
}

.grid, .ground, .buildings, .markers {
  position: absolute;
}

.tile {
  position: absolute;
}

.ground-tile, .building-tile, .marker-tile {
  width: 100%;
  height: 100%;
}

.grid {
  position: absolute;
}

.grid-tile {
  fill: none;
  stroke: gray;
  stroke-width: 1px;
}

.dot, .dot-background {
  border-radius: 50%;
  transform-origin: 50% 50%;
  width: 24px;
  height: 24px;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  opacity: 0.5;
  z-index: 2;
}

.dot {
  background: white;
}

.dot:hover {
  opacity: 1;
  transition: .3s ease-in;
  cursor: pointer;
}

.dot-background {
  opacity: 0;
  border: 0.1em solid white;
  animation: pulse 3s 2s infinite;
}

.marker-content {
  position: fixed;
  border-radius: 50%;
  width: 300px;
  height: 300px;
  transform-origin: 50% 50%;
  transform: scale(0);
}

.marker-content.open {
  transform-origin: 50% 50%;
  display: block;
  background: gray;
  padding: 100px;
  transform: scale(1);
  transition: all .3s ease-in;
}

.building-tile {
  transform: translateY(-50%);
  width: 100%;
  height: 100%;
}

@keyframes pulse {
  0% {
    opacity: 0.5;
    transform-origin: 0% 0%;
    transform: scale(1.3) translate(-50%, -50%);
  }
  100% {
    opacity: 0;
    transform-origin: 0% 0%;
    transform: scale(1.6) translate(-50%, -50%);
  }
}

.powerline, .powerflow {
  fill: none;
  opacity: 0.6;
}

.powerline {
  stroke: darkgray;
  stroke-width: 3;
}

.powerflow {
  stroke: white;
  stroke-width: 2;
  stroke-linecap: round;
  stroke-dasharray: 1 10;
  animation: powerflow 15s infinite linear forwards;
}

@keyframes powerflow {
  0% {
    stroke-dashoffset: 0;
  }

  100% {
    stroke-dashoffset: -250;
  }
}
