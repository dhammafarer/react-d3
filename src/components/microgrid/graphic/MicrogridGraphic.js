import React from 'react';
import PropTypes from 'prop-types';
import './MicrogridGraphic.scss';
import { isometricTileCoords, isometricTilePolygonPoints } from '../../../helpers/isometric-grid.js';
import TerrainTiles from './TerrainTiles.js';
import BuildingTiles from './BuildingTiles.js';
import IsometricGrid from './IsometricGrid.js';
import NetworkLines from './NetworkLines.js';
import BuildingMarkers from './BuildingMarkers.js';
import { TimelineMax } from 'gsap';

const tileRatio = 1.7345;

class MicrogridGraphic extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      size: {
        width: 400,
        height: 400 * tileRatio,
        margin: [0, 0]
      },
      tile: {
        width: 200,
        height: 200 * tileRatio
      },
      tl: new TimelineMax()
    };
  }
  componentDidMount () {
    window.addEventListener('resize', () => this.setGraphicSize());
    this.setGraphicSize();
    this.animateEnter();
  }

  componentDidUpdate (prevProps) {
    if (prevProps.gridSize != this.props.gridSize) {
      this.setGraphicSize();
    }
    if (prevProps.name != this.props.name) {
      this.animateEnter();
    }
  }

  animateEnter () {
    this.state.tl
      .clear()
      .from('.terrain-tile', 0.5, {opacity: 0.8, transform: 'scale(0.0)', y: '+=2', ease: 'Cubic.easeOut'}, 0.1)
      .from('.grid', 0.3, {opacity: 0})
      .staggerFrom('.building-tile', 0.3, {opacity: 0}, 0.2)
      .from(['.marker-tile'], 0.5, {transform: 'scale(0)', opacity: 0}, '-=0.1')
      .from('.powerline', 0.5, {opacity: 0}, "-=0.3")
      .from('.powerflow', 0.5, {opacity: 0}, "+=0.3");
  }

  setGraphicSize () {
    let componentWidth = this.graphic.offsetWidth;
    let tileWidth = componentWidth / (this.props.gridSize[1] + 1);
    let tileHeight = tileWidth / tileRatio;
    let margin = [tileHeight / 2, tileWidth / 2];
    let width = componentWidth - margin[1] * 2;
    let height = tileHeight * (this.props.gridSize[1]);
    this.setState({
      size: {width, height, margin},
      tile: {width: tileWidth, height: tileHeight}
    });
  }

  mapToIso (isotable, tile, objectsMap) {
    return objectsMap
      .map((row, y) => row
        .map((texture, x) => {
          let coords = isotable[y][x];
          let style = {
            left: coords.x,
            top: coords.y,
            width: tile.width,
            height: tile.height * texture.height
          };
          return {texture, pos: [x, y], style};
        })
      )
      .reduce((a, b) => a.concat(b), []);
  }

  arrayToIso (isotable, tile, array) {
    return array.map(el => {
      let coords = isotable[el.pos[0]][el.pos[1]];
      let style = {
        left: coords.x,
        top: coords.y - (tile.height * el.texture.offsetHeight),
        width: tile.width,
        height: tile.height * el.texture.height
      };
      el.style = style;
      return el;
    });
  }

  tileCoords (gridSize, tile) {
    let {width, height} = tile;
    return isometricTileCoords(gridSize, [width, height]);
  }

  tilePolygons (isotable, tile) {
    let {width, height} = tile;
    return isotable.map((row, y) => row
      .map((coords, x) => {
        return {
          points: isometricTilePolygonPoints([width, height], coords),
          pos: [y, x]};
      }))
      .reduce((a,b) => a.concat(b), []);
  }

  render () {
    let {width, height, margin} = this.state.size;
    let tile = this.state.tile;
    let {gridSize, terrainMap, buildings} = this.props;
    let graphicStyles = {
      width, height, margin: margin[0] + 'px ' + margin[1] + 'px'
    };

    let isotable = this.tileCoords(gridSize, tile);
    let terrainLayer = this.mapToIso(isotable, tile, terrainMap).filter(el => el.texture);
    let buildingsLayer = this.arrayToIso(isotable, tile, buildings);
    let grid = this.tilePolygons(isotable, tile);

    return (
      <div className="system-graphic" ref={graphic => this.graphic = graphic}>
        <h2>{this.props.name}</h2>
        <div className="graphic-content" style={graphicStyles}>

          <TerrainTiles terrain={terrainLayer}/>
          <IsometricGrid grid={grid} width={width} height={height}/>
          <BuildingTiles buildings={buildingsLayer}/>
          <NetworkLines data={buildings} tile={this.state.tile} width={width} height={height}/>
          <BuildingMarkers data={buildings} handleClick={this.props.openGraphicModal}/>

        </div>
      </div>
    );
  }
}

MicrogridGraphic.propTypes = {
  name: PropTypes.string.isRequired,
  gridSize: PropTypes.array.isRequired,
  terrainMap: PropTypes.array.isRequired,
  buildings: PropTypes.array.isRequired,
  openGraphicModal: PropTypes.func.isRequired

};

export default MicrogridGraphic;
