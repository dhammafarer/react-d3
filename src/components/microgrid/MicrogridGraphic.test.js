/* globals describe, it, expect*/
import React from 'react';
import renderer from 'react-test-renderer';
import MicrogridGraphic from './MicrogridGraphic.js';

describe('MicrogridGraphic', () => {
  it('displays the graphic', () => {
    const options = {createNodeMock};
    const tree = renderer.create(
      <MicrogridGraphic />,
      options
    );

    expect(tree).toMatchSnapshot();
  });
});

function createNodeMock (element) {
  if (element.type === 'div') {
    return {style: {}, offsetWidth: 800};
  }
  return null;
}
