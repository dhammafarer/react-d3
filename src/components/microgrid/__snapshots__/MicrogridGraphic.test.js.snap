// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MicrogridGraphic displays the graphic 1`] = `
<div
  className="system-graphic"
>
  <div
    className="graphic-content"
    style={
      Object {
        "height": 345.9210147016431,
        "margin": "57.653502450273855px 100px",
        "width": 600,
      }
    }
  >
    <div
      className="ground"
    >
      <div
        className="tile"
        style={
          Object {
            "left": 200,
            "top": 0,
            "width": 200,
          }
        }
      >
        <img
          className="ground-tile"
          src="grass.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 300,
            "top": 57.653502450273855,
            "width": 200,
          }
        }
      >
        <img
          className="ground-tile"
          src="grass.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 400,
            "top": 115.30700490054771,
            "width": 200,
          }
        }
      >
        <img
          className="ground-tile"
          src="grass.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 100,
            "top": 57.653502450273855,
            "width": 200,
          }
        }
      >
        <img
          className="ground-tile"
          src="grass.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 200,
            "top": 115.30700490054771,
            "width": 200,
          }
        }
      >
        <img
          className="ground-tile"
          src="grass.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 300,
            "top": 172.96050735082156,
            "width": 200,
          }
        }
      >
        <img
          className="ground-tile"
          src="grass.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 0,
            "top": 115.30700490054771,
            "width": 200,
          }
        }
      >
        <img
          className="ground-tile"
          src="grass.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 100,
            "top": 172.96050735082156,
            "width": 200,
          }
        }
      >
        <img
          className="ground-tile"
          src="grass.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 200,
            "top": 230.61400980109542,
            "width": 200,
          }
        }
      >
        <img
          className="ground-tile"
          src="grass.png"
        />
      </div>
    </div>
    <div
      className="buildings"
    >
      <div
        className="tile"
        style={
          Object {
            "left": 200,
            "top": 0,
            "width": 200,
          }
        }
      >
        <img
          className="building-tile"
          src="temple.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 400,
            "top": 115.30700490054771,
            "width": 200,
          }
        }
      >
        <img
          className="building-tile"
          src="solar.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 100,
            "top": 57.653502450273855,
            "width": 200,
          }
        }
      >
        <img
          className="building-tile"
          src="house.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 300,
            "top": 172.96050735082156,
            "width": 200,
          }
        }
      >
        <img
          className="building-tile"
          src="house.png"
        />
      </div>
      <div
        className="tile"
        style={
          Object {
            "left": 100,
            "top": 172.96050735082156,
            "width": 200,
          }
        }
      >
        <img
          className="building-tile"
          src="factory.png"
        />
      </div>
    </div>
    <svg
      className="grid"
      height={345.9210147016431}
      width={600}
    >
      <polygon
        className="grid-tile"
        data-pos={
          Array [
            0,
            0,
          ]
        }
        points="300,0 400,57.653502450273855 300,115.30700490054771 200,57.653502450273855"
      />
      <polygon
        className="grid-tile"
        data-pos={
          Array [
            0,
            1,
          ]
        }
        points="400,57.653502450273855 500,115.30700490054771 400,172.96050735082156 300,115.30700490054771"
      />
      <polygon
        className="grid-tile"
        data-pos={
          Array [
            0,
            2,
          ]
        }
        points="500,115.30700490054771 600,172.96050735082156 500,230.61400980109542 400,172.96050735082156"
      />
      <polygon
        className="grid-tile"
        data-pos={
          Array [
            1,
            0,
          ]
        }
        points="200,57.653502450273855 300,115.30700490054771 200,172.96050735082156 100,115.30700490054771"
      />
      <polygon
        className="grid-tile"
        data-pos={
          Array [
            1,
            1,
          ]
        }
        points="300,115.30700490054771 400,172.96050735082156 300,230.61400980109542 200,172.96050735082156"
      />
      <polygon
        className="grid-tile"
        data-pos={
          Array [
            1,
            2,
          ]
        }
        points="400,172.96050735082156 500,230.61400980109542 400,288.2675122513693 300,230.61400980109542"
      />
      <polygon
        className="grid-tile"
        data-pos={
          Array [
            2,
            0,
          ]
        }
        points="100,115.30700490054771 200,172.96050735082156 100,230.61400980109542 0,172.96050735082156"
      />
      <polygon
        className="grid-tile"
        data-pos={
          Array [
            2,
            1,
          ]
        }
        points="200,172.96050735082156 300,230.61400980109542 200,288.2675122513693 100,230.61400980109542"
      />
      <polygon
        className="grid-tile"
        data-pos={
          Array [
            2,
            2,
          ]
        }
        points="300,230.61400980109542 400,288.2675122513693 300,345.9210147016431 200,288.2675122513693"
      />
    </svg>
  </div>
</div>
`;
